Add detailed steps for constructing RTCIceCandidate

soareschen opened the pull request:
This PR address the issues in #1165 and #1224. It is also needed to have complete test coverage for `RTCIceCandidate()` in w3c/web-platform-tests#5917.

I am not very sure of the details for editing the spec, so this is a first cut draft that attempt to explain the changes to be made.

The main change is to add detailed steps for the `RTCIceCandidate` constructor to parse the candidate string, referencing the relevant specs, and populating the attributes with the parse result. 

To maintain backward compatibility, any parsing error is ignored and the derivative attributes are simply set with the default `null` value. This might not be the best approach, but I am not sure if we can make it throw an error since current browser implementations do not have the derivative attributes and have not done any parsing on the candidate string.

taylor-b commented on webrtc.html:
```diff
@@ -3835,18 +3835,49 @@
+                  <code>RTCIceCandidate</code> object.</li>
+                  <li>Set the corresponding attributes in <var>iceCandidate</var>
+                  with the dictionary member values of <var>candidateInitDict</var></li>
+                  <li>Let <var>candidate</var> be the <code><a>candidate</a></code>
```
Does this `<a>` link to the right `candidate`?

taylor-b commented on webrtc.html:
```diff
@@ -3835,18 +3835,49 @@
+                  is not done in the constructor for <code>RTCIceCandidate</code>.
+                  Instead, validation is done when calling <code><a data-for=
+                  "RTCPeerConnection">addIceCandidate()</a></code>. To maintain
+                  backward compatibility, if the <var>candidate</var> member do not
```
s/do/does

taylor-b commented on webrtc.html:
```diff
@@ -3835,18 +3835,49 @@
+                  backward compatibility, if the <var>candidate</var> member do not
+                  match the grammar for <code>candidate-attribute</code>, the value
+                  remain set in the <code>candidate</code> attribute with the other
+                  derivative fields set to <code>null</code>
```
Meaning, the DOMString `candidate` attribute will return the string passed in, but the parsed attributes will return `null`? I think this could be more clear.

taylor-b commented on webrtc.html:
```diff
@@ -3835,18 +3835,49 @@
+                    defined in section 4.5 of [[!RFC6544]].</li>
+                    <li>If parsing of <code>candidate-attribute</code> is successful,
+                    set the corresponding attributes in <var>iceCandidate</var> to
+                    the field values of the parse result.</li>
```
s/parse/parsed?

taylor-b commented on webrtc.html:
```diff
@@ -3835,18 +3835,49 @@
+                </ol>
+                <div class="note">
+                  The validation for members in <var>candidateInitDict</var>
+                  is not done in the constructor for <code>RTCIceCandidate</code>.
```
This statement is somewhat misleading, since `sdpMid` and `sdpMLineIndex` (or at least, their presence) are validated above.

taylor-b commented on webrtc.html:
```diff
@@ -3835,18 +3835,49 @@
+                  <li>Let <var>iceCandidate</var> be a newly created
+                  <code>RTCIceCandidate</code> object.</li>
+                  <li>Set the corresponding attributes in <var>iceCandidate</var>
+                  with the dictionary member values of <var>candidateInitDict</var></li>
```
Missing period

taylor-b commented on webrtc.html:
```diff
@@ -3835,18 +3835,49 @@
+                    set the corresponding attributes in <var>iceCandidate</var> to
+                    the field values of the parse result.</li>
+                  </ol></li>
+                  <li>Return <var>iceCandidate</var></li>
```
Missing period.

taylor-b commented on webrtc.html:
```diff
@@ -3835,18 +3835,49 @@
+                  <var>candidate</var> is not an empty string, run the following steps:
+                  <ol>
+                    <li>Parse <var>candidate</var> using the grammar for
+                    <code>candidate-attribute</code> as defined in section 15.1 of
```
Can you use `<dfn>` for `candidate-attribute` and link to the definition where it's referenced below?

soareschen commented:
> Another question: what should happen if a field has a value that doesn't match any corresponding enum value. For example, a type that isn't "host/srflx/relay", but something else, possibly from an ICE extension the browser doesn't support.

This seems to related to #1211. We'd then have to relax the constraint and allow the `candidate-attribute` grammar to be extensible.

I have updated the paragraphs and attempt to clarify things based on feedback.

soareschen commented:
Rebased. I have added reference to `candidate-attribute` for the new `component` field in `RTCIceCandidate`. I also clarify the parsing step to require the parsing to either all pass or fail. e.g. If `component-id` has a value other than `1` or `2`, or if the IP address is malformed, the steps are aborted and all derivative attributes are set to null even if the other field values are valid. This is to prevent browsers from doing non strict parsing which could make it difficult to interpret the fields.

Please let me know if the proposal is sound. Frankly I feel that the simpler approach is to make all attributes non-nullable and simply throw an error when parsing failed. Because otherwise it would be hard for users to use these fields if their values are nullable depending on whether parsing is successful.

taylor-b commented on webrtc.html:
```diff
@@ -3839,18 +3844,60 @@
+                </ol>
+                <div class="note">
+                  <p>The constructor for <code>RTCIceCandidate</code> only does basic
+                  type checking for the dictionary members in <var>candidateInitDict</var>.
```
This sentence seems slightly misleading, since it doesn't just do basic type checking; it also does basic parsing.

soareschen replied:
I have updated the sentence to _"only does basic parsing and type checking"_.

taylor-b commented on webrtc.html:
```diff
@@ -3839,18 +3844,60 @@
+                  <code>candidate</code></a> string given in <var>candidateInitDict</var>,
+                  but derivative attributes such as <a><code>foundation</code></a>,
+                  <a><code>priority</code></a>, etc are set to their default value,
+                  which is <code>null</code>.</p>
```
The "derivative attributes are set to `null` part" could be a bit more clear in the normative text. Maybe step 1 could be "initialize attribute values to `null`". Or is this implied?

soareschen replied:
I think the default value for nullable attributes is `null`. Anyway I clarified it in the constructor steps by explicitly listing the attributes to initialize.

stefhak commented:
Is there any conflict between this one and https://github.com/w3c/webrtc-pc/pull/1372/files? I don't think so, and I think this one can be merged!
